modules/tests/src/test/scala/dev/profunktor/redis4cats/TestScenarios.scala

```diff
@@ -575,4 +578,49 @@ trait TestScenarios { self: FunSuite =>
       _ <- IO(assert(c4 > 0, "merged hyperloglog should think it has more than 0 items in"))
     } yield ()
   }
+
+  def keyPatternSubScenario(client: RedisClient): IO[Unit] = {
+    import dev.profunktor.redis4cats.effect.Log.NoOp._
+    import scala.language.postfixOps
+
+    val pattern = "__keyevent*__:*"
+    val key     = "somekey"
+    val resources = for {
+      ref <- Resource.eval(Ref.of[IO, Option[RedisPatternEvent[String, String]]](None))
+      commands <- Redis[IO].fromClient(client, RedisCodec.Utf8)
+      sub <- PubSub.mkSubscriberConnection[IO, String, String](client, RedisCodec.Utf8)
+      stream <- Resource.pure(sub.psubscribe(RedisPattern(pattern)))
+      _ <- stream.foreach(output => ref.set(Some(output))).compile.drain.background
+      _ <- Resource.eval(commands.set(key, ""))
+      _ <- Resource.eval(commands.expire(key, 1 seconds))
+    } yield ref.get
+    resources.use(get => get.iterateUntil(_.isDefined).timeout(2 seconds)).map { result =>
+      assert(
+        result == Some(RedisPatternEvent(pattern, "__keyevent@0__:expired", key)),
+        s"Unexpected result $result"
+      )
+    }
+  }
+
+  def channelPatternSubScenario(client: RedisClient): IO[Unit] = {
+    import dev.profunktor.redis4cats.effect.Log.NoOp._
+    import scala.language.postfixOps
+
+    val pattern = "f*"
+    val channel = "foo"
+    val message = "somemessage"
+    val resources = for {
+      ref <- Resource.eval(Ref.of[IO, Option[RedisPatternEvent[String, String]]](None))
+      pubsub <- PubSub.mkPubSubConnection[IO, String, String](client, RedisCodec.Utf8)
+      stream <- Resource.pure(pubsub.psubscribe(RedisPattern(pattern)))
+      _ <- stream.foreach(output => ref.set(Some(output))).compile.drain.background
+      _ <- Resource.eval(pubsub.publish(RedisChannel(channel))(fs2.Stream(message)).compile.drain)
+    } yield ref.get
+    resources.use(get => get.iterateUntil(_.isDefined).timeout(2 seconds)).map { result =>
```

Comment:
Not sure what's the underlying type, but could this perhaps be `nonEmpty` instead of `isDefined`? Just syntactic meaningfulness.

Reply:
Refactored it :-)